### PR TITLE
Add a new blog: how to trace linux syscalls

### PR DESCRIPTION
**Pull request by TomShawn**

### CODE REVIEW

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
++ sched: Analyzes scheduler actions and latencies.
++ timechart: Visualizes system behaviors based on the workload.
++ c2c: Detects the potential for false sharing. Red Hat once tested the c2c prototype on a number of Linux applications and found many cases of false sharing and cache lines on hotspots.
++ trace: Traces system calls with acceptable overheads. It performs only **1.36** times slower with workloads specified in the `dd` command.
+
+Let’s look at some common uses of perf.
+
++ To see which commands made the most system calls:
+
+    ```shell
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:29]: Please add copyable code to all code blocks.

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    mkdir /sys/fs/cgroup/perf_event/bpftools/
+    echo 22542 >> /sys/fs/cgroup/perf_event/bpftools/tasks
+    echo 20514 >> /sys/fs/cgroup/perf_event/bpftools/tasks
+    perf trace -G bpftools -a -- sleep 10
+    ```
+
+    ![*Trace a group of tasks*](media/trace-a-group-of-tasks.jpg)
+
+    Those are some of the most common uses of perf.  If you’d like to know more (especially about perf-trace), see the [Linux manual page](https://man7.org/linux/man-pages/man1/perf-trace.1.html). From the manual pages, you will learn that perf-trace can filter tasks based on PIDs or thread IDs (TIDs), but that it has no convenient support for containers and the Kubernetes (K8s) environments. Don’t worry. Next, we’ll discuss a tool that can easily trace system calls in containers and in K8s environments that uses cgroup v2.
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:76]:
```suggestion
Those are some of the most common uses of perf.  If you’d like to know more (especially about perf-trace), see the [Linux manual page](https://man7.org/linux/man-pages/man1/perf-trace.1.html). From the manual pages, you will learn that perf-trace can filter tasks based on PIDs or thread IDs (TIDs), but that it has no convenient support for containers and the Kubernetes (K8s) environments. Don’t worry. Next, we’ll discuss a tool that can easily trace system calls in containers and in K8s environments that uses cgroup v2.
```

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+We can further simplify this illustration into the following key procedures. Note that these procedures are implementation details, not operations to perform:
+
+1. `bpf helper` gets the cgroup ID. Tasks are filtered based on the cgroup ID rather than on the PID and TID.
+2. Each cgroup ID corresponds to a [bpf tail call](https://ebpf.io/what-is-ebpf/#tail--function-calls) that can call and execute another eBPF program and replace the execution context. Syscall events are written through a bpf tail call to a perf ring buffer with the same cgroup ID.
+3. The user space reads the perf ring buffer based on this cgroup ID.
+
+> **Note:**
+>
+> Currently, you  can get the cgroup ID only by executing `bpf helper: bpf_get_current_cgroup_id`, and this ID is available only in cgroup v2. Therefore, before you use traceloop, make sure that [cgroup v2 is enabled](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#basic-operations) in your environment.
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:92]:
```suggestion
> Currently, you can get the cgroup ID only by executing `bpf helper: bpf_get_current_cgroup_id`, and this ID is available only in cgroup v2. Therefore, before you use traceloop, make sure that [cgroup v2 is enabled](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#basic-operations) in your environment.
```

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+As the results show, the traceloop output is similar to that of strace or perf-trace except for the cgroup-based task filtering. Note that CentOS 8 mounts cgroup v2 directly on the `/sys/fs/cgroup` path instead of on `/sys/fs/cgroup/unified` as Ubuntu does. Therefore, before you use traceloop, you should run `mount -t cgroup2` to determine the mount information.
+
+The team behind traceloop has integrated it with the Inspektor Gadget project, so you can run traceloop on the K8s platform using kubectl. See the demos in [Inspektor Gadget - How to use](https://github.com/kinvolk/inspektor-gadget#how-to-use) and, if you like, try it on your own.
+
+## Benchmark with system calls traced
+
+We conducted a sysbench test in which system calls were either traced using multiple tracers (traceloop, strace, and perf-trace) or not traced. The benchmark results are as follows:
+
+![*Sysbench results with system calls traced and untraced*](media/sysbench-results-with-system-calls-traced-and-untraced.jpg)
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md]: Please add an image caption.

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md (hunk 2 above)
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:76]:
```suggestion
    Those are some of the most common uses of perf. If you’d like to know more (especially about perf-trace), see the [Linux manual page](https://man7.org/linux/man-pages/man1/perf-trace.1.html). From the manual pages, you will learn that perf-trace can filter tasks based on PIDs or thread IDs (TIDs), but that it has no convenient support for containers and the Kubernetes (K8s) environments. Don’t worry. Next, we’ll discuss a tool that can easily trace system calls in containers and in K8s environments that uses cgroup v2.
```

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md (hunk 3 above)
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:92]:
```suggestion
> Currently, you can get the cgroup ID only by executing `bpf helper: bpf_get_current_cgroup_id`, and this ID is available only in cgroup v2. Therefore, before you use traceloop, make sure that [cgroup v2 is enabled](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#basic-operations) in your environment.
```

Comment (CaitinChen) [reply how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:92]: If this note is for bullet item #3, you must indent it.

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Traceloop, a performance profiler for cgroup v2 and K8s
+
+Traceloop provides better support for tracing Linux system calls in the containers or K8s environments that use cgroup v2. You might be unfamiliar with traceloop but know BPF Compiler Collection (BCC) pretty well. (Its front-end is implemented using Python or C++.) In the IO Visor Project, BCC’s parent project, there is another project named gobpf that provides Golang bindings for the BCC framework. Based on gobpf, traceloop is developed for environments of containers and K8s. The following illustration shows the traceloop architecture:
+
+![*traceloop architecture*](media/traceloop-architecture.jpg)
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md]: Please add an image caption.

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md (hunk 3 above)
Comment (TomShawn) [reply how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md]: No, the note is for the whole list.

**Comment by TomShawn** [comment]
@CaitinChen All comments are addressed, PTAL again, thanks!

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+We conducted a sysbench test in which system calls were either traced using multiple tracers (traceloop, strace, and perf-trace) or not traced. The benchmark results are as follows:
+
+![Sysbench results with system calls traced and untraced](media/sysbench-results-with-system-calls-traced-and-untraced.jpg)
+<div class="caption-center"> Sysbench results with system calls traced and untraced </div>
+
+As the benchmark shows, strace caused the biggest decrease in application performance. perf-trace caused a smaller decrease, and traceloop caused the smallest.
+
+## Summary of Linux profilers
+
+For issues such as "Why can't the software run on this machine," strace is still a powerful system call tracer in Linux. But to trace the latency of system calls, the BPF-based perf-trace is a better option. In containers or K8s environments that use cgroup v2, traceloop is the easiest to use.
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:137]: Please add a line after this paragraph.

File: how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+In the following demo (on the CentOS 8 4.18 kernel), when traceloop exits, the system call information is traced:
+
+{{< copyable "shell-regular" >}}
+
+```shell
+sudo -E ./traceloop cgroups --dump-on-exit /sys/fs/cgroup/system.slice/sshd.service
+```
+
+![traceloop tracing system calls](media/traceloop-tracing-system-calls.jpg)
+<div class="caption-center"> traceloop tracing system calls </div>
```
Comment (CaitinChen) [inline how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md:120]:
```suggestion
<div class="caption-center"> Traceloop tracing system calls </div>
```

**Review by CaitinChen — APPROVED** [review]

**Pull request merged**